File: src/main/python/AoC2025_07.py
#! /usr/bin/env python3
#
# Advent of Code 2025 Day 7
#

import sys

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples

Input = tuple[str, ...]
Output1 = int
Output2 = int


TEST = """\
.......S.......
...............
.......^.......
...............
......^.^......
...............
.....^.^.^.....
...............
....^.^...^....
...............
...^.^...^.^...
...............
..^...^.....^..
...............
.^.^.^.^.^...^.
...............
"""

SPLITTER = "^"
START = "S"


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return tuple(input_data)

    def solve(self, grid: Input) -> tuple[int, int]:
        start = grid[0].index(START)
        beams = [0] * len(grid[0])
        beams[start] = 1
        splits = 0
        for r in range(1, len(grid)):
            new_beams = [0] * len(grid[0])
            for c in range(len(grid[0])):
                if grid[r][c] == SPLITTER:
                    new_beams[c - 1] += beams[c]
                    new_beams[c + 1] += beams[c]
                    splits += beams[c] > 0
                else:
                    new_beams[c] += beams[c]
            beams = new_beams
        return splits, sum(beams)

    def part_1(self, grid: Input) -> Output1:
        return self.solve(grid)[0]

    def part_2(self, grid: Input) -> Output2:
        return self.solve(grid)[1]

    @aoc_samples(
        (
            ("part_1", TEST, 21),
            ("part_2", TEST, 40),
        )
    )
    def samples(self) -> None:
        pass


solution = Solution(2025, 7)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/java/AoC2025_07.java
import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

import java.util.Arrays;
import java.util.List;

@SuppressWarnings({"PMD.ClassNamingConventions", "PMD.NoPackage"})
public final class AoC2025_07 extends SolutionBase<List<String>, Long, Long> {

    public static final char SPLITTER = '^';
    public static final char START = 'S';

    private AoC2025_07(final boolean debug) {
        super(debug);
    }

    public static AoC2025_07 create() {
        return new AoC2025_07(false);
    }

    public static AoC2025_07 createDebug() {
        return new AoC2025_07(true);
    }

    @Override
    protected List<String> parseInput(final List<String> inputs) {
        return inputs;
    }

    @SuppressWarnings("PMD.AvoidInstantiatingObjectsInLoops")
    private Result solve(final List<String> grid) {
        final int start = grid.getFirst().indexOf(START);
        long[] beams = new long[grid.getFirst().length()];
        beams[start] = 1;
        long splits = 0;
        for (int r = 1; r < grid.size(); r++) {
            final char[] row = grid.get(r).toCharArray();
            final long[] newBeams = new long[row.length];
            for (int c = 0; c < row.length; c++) {
                if (row[c] == SPLITTER) {
                    newBeams[c - 1] += beams[c];
                    newBeams[c + 1] += beams[c];
                    splits += (beams[c] > 0 ? 1 : 0);
                } else {
                    newBeams[c] += beams[c];
                }
            }
            beams = newBeams;
        }
        return new Result(splits, Arrays.stream(beams).sum());
    }

    @Override
    public Long solvePart1(final List<String> grid) {
        return this.solve(grid).splits;
    }

    @Override
    public Long solvePart2(final List<String> grid) {
        return this.solve(grid).beams;
    }

    @Samples({
        @Sample(method = "part1", input = TEST, expected = "21"),
        @Sample(method = "part2", input = TEST, expected = "40"),
    })
    public static void main(final String[] args) throws Exception {
        create().run();
    }

    private static final String TEST =
            """
            .......S.......
            ...............
            .......^.......
            ...............
            ......^.^......
            ...............
            .....^.^.^.....
            ...............
            ....^.^...^....
            ...............
            ...^.^...^.^...
            ...............
            ..^...^.....^..
            ...............
            .^.^.^.^.^...^.
            ...............
            """;

    private record Result(long splits, long beams) {}
}
